### PR TITLE
Re-add exact choice shapeless support

### LINKED ISSUE #11593: RecipeBook click ignores ExactChoice
### Expected behavior

Custom recipe to be selected correctly.

### Observed/Actual behavior

If you click on the recipe button you clearly see vanilla items added instead of the renamed Diamond.
This causes the recipe to not be selected correctly, instead the vanilla overlapping recipe gets used.

This happens both on Spigot (even on pre-1.21.3) and on Paper 1.21.3.
Paper 1.21.1 [implemented a fix](https://github.com/PaperMC/Paper/blob/ver/1.21.1/patches/server/0277-Improve-exact-choice-recipe-ingredients.patch) for that which wasn't introduced in 1.21.3.

I did some tests by logging client packets, using [this guide](https://wiki.vg/Debugging), and seems that 1.21.3 client doesn't send `[play:serverbound/minecraft:place_recipe]` (`ServerboundPlaceRecipePacket`) packet at all when clicking again on the recipe. It used to do that on 1.21.1, so this might now be fixable.

### Steps/models to reproduce

# 1.21.1
https://github.com/user-attachments/assets/b0cb5ba1-6503-4948-9f03-e8398100ece8

# 1.21.3
https://github.com/user-attachments/assets/5b3c61d5-236c-447c-8846-6abd77e8a299

Example code to reproduce the issue:

```java
Player player = Bukkit.getPlayer("LoneDev");
org.bukkit.inventory.ShapedRecipe recipe = new org.bukkit.inventory.ShapedRecipe(new NamespacedKey("test", "test_exact_choice"), new org.bukkit.inventory.ItemStack(Material.BARRIER));
recipe.shape( "ccc",  "xix",  "xix");
org.bukkit.inventory.ItemStack exact = new org.bukkit.inventory.ItemStack(Material.DIAMOND);
ItemMeta meta = exact.getItemMeta();
meta.setDisplayName("test item");
exact.setItemMeta(meta);
recipe.setIngredient('c', new RecipeChoice.ExactChoice(exact));
recipe.setIngredient('i', new RecipeChoice.MaterialChoice(Material.STICK));

org.bukkit.inventory.ItemStack aaa = exact.clone();
aaa.setAmount(32);
player.getInventory().addItem(aaa);

if(Bukkit.getRecipe(recipe.getKey()) != null)
    Bukkit.removeRecipe(recipe.getKey());
Bukkit.addRecipe(recipe); 
```

### Plugin and Datapack List

None

### Paper version

This server is running Paper version 1.21.3-14-master@fcb6c72 (2024-11-06T10:06:32Z) (Implementing API version 1.21.3-R0.1-SNAPSHOT)
You are running the latest version

### Other

Opened an issue on Spigot jira too as this bug is reproducible also on Spigot.
https://hub.spigotmc.org/jira/browse/SPIGOT-7950

### PR DESCRIPTION
**Pull request by Machine-Maker**

### CODE REVIEW

**Comment by Machine-Maker** [comment]
Tested with exact choice shapeless recipes. As always, there are a few caveats with how these work in relation to the client. If your exact choice recipe doesn't include ingredients with damaged items, enchanted items, or "custom_name" items (NOT item_name), then they should work 100%. If you have ingredients that are damaged, enchanted, or have "custom_name" (displayName) NOT "item_name" (itemName), then the recipe book will show them as uncraftable and multiple clicks on the item won't keep adding 1 "recipe" to the grid, you get one click (or shift click).

File: patches/server/1058-Improve-exact-choice-recipe-ingredients.patch [hunk 1]
```diff
@@ -0,0 +1,432 @@
...
++            this.contents.amounts.removeInt(new ItemOrExact.Exact(extra));
++        }
++        for (final Object2IntMap.Entry<ItemOrExact.Item> entry : this.regularRemoved.object2IntEntrySet()) {
++            this.contents.amounts.addTo(entry.getKey(), entry.getIntValue());
++        }
++    }
++
++    public boolean accountStack(final ItemStack stack, final int count) {
++        if (!this.exactIngredients.isEmpty()) {
++            if (this.exactIngredients.contains(stack)) {
```
Comment (kennytv) [inline patches/server/1058-Improve-exact-choice-recipe-ingredients.patch]: I'd merge these into one

File: patches/server/1058-Improve-exact-choice-recipe-ingredients.patch [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
++                                io.papermc.paper.inventory.recipe.ItemOrExact holder = list.get(jx); // Paper - Improve exact choice recipe ingredients
+                                 kx = this.moveItemToGrid(slot2, holder, kx);
+                                 if (kx == -1) {
+                                     return;
+@@ -155,7 +156,7 @@ public class ServerPlaceRecipe<R extends Recipe<?>> {
+         }
+     }
+ 
+-    private int moveItemToGrid(Slot slot, Holder<Item> item, int count) {
++    private int moveItemToGrid(Slot slot, io.papermc.paper.inventory.recipe.ItemOrExact item, int count) {
```
Comment (kennytv) [inline patches/server/1058-Improve-exact-choice-recipe-ingredients.patch]: Comment missing

File: patches/server/1058-Improve-exact-choice-recipe-ingredients.patch [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+diff --git a/src/main/java/net/minecraft/world/entity/player/StackedContents.java b/src/main/java/net/minecraft/world/entity/player/StackedContents.java
+index 3bc7c9b3df5547c6564696fce67417c96ffdebf9..1bc00550b5a31beaa44b9f83b8a6eefea3a36c73 100644
+--- a/src/main/java/net/minecraft/world/entity/player/StackedContents.java
++++ b/src/main/java/net/minecraft/world/entity/player/StackedContents.java
+@@ -10,7 +10,7 @@ import java.util.Set;
+ import javax.annotation.Nullable;
+ 
+ public class StackedContents<T> {
+-    public final Reference2IntOpenHashMap<T> amounts = new Reference2IntOpenHashMap<>();
++    public final it.unimi.dsi.fastutil.objects.Object2IntOpenHashMap<T> amounts = new it.unimi.dsi.fastutil.objects.Object2IntOpenHashMap<>();
```
Comment (kennytv) [inline patches/server/1058-Improve-exact-choice-recipe-ingredients.patch]: Also here

**Review by kennytv — APPROVED** [review]

**Review by lynxplay — APPROVED** [review]

**Pull request merged**